src/persist/src/indexed/mod.rs

```diff
@@ -0,0 +1,462 @@
+                .futures
+                .iter()
+                .map(|(id, future)| (id.0, future.meta()))
+                .collect(),
+            traces: self
+                .traces
+                .iter()
+                .map(|(id, trace)| (id.0, trace.meta()))
+                .collect(),
+        }
+    }
+
+    /// Returns a [Snapshot] for the given id.
+    pub fn snapshot(&self, id: Id) -> Result<IndexedSnapshot, Error> {
+        let mut buffer = BufferSnapshot(Vec::new());
+        let buf_lower = {
+            self.buf
+                .snapshot(|_, buf| {
+                    let entry: Abomonated<BufferEntry, Vec<u8>> =
+                        unsafe { Abomonated::new(buf.to_owned()) }
+                            .ok_or_else(|| Error::from(format!("invalid buffer entry")))?;
+                    if entry.id != id.0 {
+                        return Ok(());
+                    }
+                    buffer.0.extend(entry.updates.iter().cloned());
+                    Ok(())
+                })?
+                .start
+        };
+
+        let future = self
+            .futures
+            .get(&id)
+            .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
+        let trace = self
+            .traces
+            .get(&id)
+            .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
+        let trace = trace.snapshot(&self.blob)?;
+        let future = future.snapshot(buf_lower, trace.frontier.clone(), &self.blob)?;
+        Ok(IndexedSnapshot(buffer, future, trace))
+    }
+}
+
+/// A consistent snapshot of the data currently in a [Buffer].
+struct BufferSnapshot(Vec<((String, String), u64, isize)>);
+
+impl Snapshot for BufferSnapshot {
+    fn read<E: Extend<((String, String), u64, isize)>>(&mut self, buf: &mut E) -> bool {
+        buf.extend(self.0.drain(..));
+        false
+    }
+}
+
+/// A consistent snapshot of all data currently stored for an id.
+pub struct IndexedSnapshot(BufferSnapshot, FutureSnapshot, TraceSnapshot);
+
+impl Snapshot for IndexedSnapshot {
+    fn read<E: Extend<((String, String), u64, isize)>>(&mut self, buf: &mut E) -> bool {
+        self.0.read(buf) || self.1.read(buf) || self.2.read(buf)
+    }
+}
+
+/// The structure serialized and stored as an entry in a Buffer.
+#[derive(Debug, Abomonation)]
+pub struct BufferEntry {
+    /// Id of the stream this batch belongs to.
+    id: u64,
+    /// The updates themselves.
+    updates: Vec<((String, String), u64, isize)>,
+}
+
+/// The structure serialized and stored as a value in [Blob] storage for
+/// metadata keys.
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobMeta {
+    last_file_id: u128,
+    futures: Vec<(u64, BlobFutureMeta)>,
+    traces: Vec<(u64, BlobTraceMeta)>,
+}
+
+/// The metadata necessary to reconstruct a BlobFuture.
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobFutureMeta {
+    frontier: Antichain<u64>,
+    batches: Vec<((SeqNo, SeqNo), String)>,
```

Comment:
is this + the desc in `BlobFutureBatch` meant to be like a [lower bound, upper bound) of sequence numbers in this BlobFuture? I _think_ maybe its clearer to use a `Description<SeqNo>` to indicate that?